// tree-extended/dist/Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FilterConfigurationItem_1 = require("./filters/FilterConfigurationItem");
var Configuration = /** @class */ (function () {
    function Configuration(charset, maximumDeep, showIndicatorWhenDirectoryIsNotEmpty, includeGitIgnore, ignoreFilters, onlyFilters) {
        if (charset === void 0) { charset = null; }
        if (maximumDeep === void 0) { maximumDeep = null; }
        if (showIndicatorWhenDirectoryIsNotEmpty === void 0) { showIndicatorWhenDirectoryIsNotEmpty = false; }
        if (includeGitIgnore === void 0) { includeGitIgnore = false; }
        if (ignoreFilters === void 0) { ignoreFilters = ""; }
        if (onlyFilters === void 0) { onlyFilters = ""; }
        this.charset = charset;
        this.maximumDeep = maximumDeep;
        this.showIndicatorWhenDirectoryIsNotEmpty = showIndicatorWhenDirectoryIsNotEmpty;
        this.includeGitIgnore = includeGitIgnore;
        this.ignoreFilters = FilterConfigurationItem_1.default.parseArray(ignoreFilters);
        this.onlyFilters = FilterConfigurationItem_1.default.parseArray(onlyFilters);
    }
    return Configuration;
}());
exports.default = Configuration;
