// tree-extended/dist/filters/FilterConfigurationItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMPTY_DEEP = void 0;
exports.EMPTY_DEEP = -1;
var FilterConfigurationItem = /** @class */ (function () {
    function FilterConfigurationItem(pattern, deep) {
        if (deep === void 0) { deep = exports.EMPTY_DEEP; }
        this.deep = deep;
        this.pattern = pattern;
        this.regex = FilterConfigurationItem.getRegexFromPattern(pattern);
    }
    FilterConfigurationItem.prototype.isMatch = function (path, testDeep) {
        if (testDeep === void 0) { testDeep = exports.EMPTY_DEEP; }
        if (FilterConfigurationItem.deepsAreDifferentAndNotEmpty(this.deep, testDeep)) {
            return false;
        }
        return this.regex.test(path);
    };
    FilterConfigurationItem.prototype.isValid = function () {
        return this.pattern !== null && this.pattern !== undefined && !isNaN(this.deep);
    };
    FilterConfigurationItem.prototype.isEmpty = function () {
        return this.pattern === '';
    };
    FilterConfigurationItem.getRegexFromPattern = function (pattern) {
        return pattern instanceof RegExp ? pattern : new RegExp(pattern);
    };
    FilterConfigurationItem.deepsAreDifferentAndNotEmpty = function (currentDeep, testDeep) {
        return currentDeep !== exports.EMPTY_DEEP && testDeep !== exports.EMPTY_DEEP && currentDeep !== testDeep;
    };
    FilterConfigurationItem.parse = function (argItem) {
        var itemArray = argItem.trim().split(':');
        var result = null;
        if (itemArray.length === 1)
            result = new FilterConfigurationItem(itemArray[0]);
        if (itemArray.length === 2)
            result = new FilterConfigurationItem(itemArray[1], Number(itemArray[0]));
        if (result !== null && (!result.isValid() || result.isEmpty()))
            return null;
        return result;
    };
    FilterConfigurationItem.parseArray = function (arg) {
        return arg
            .split(',')
            .map(FilterConfigurationItem.parse)
            .filter(function (filterItem) { return filterItem !== null; });
    };
    return FilterConfigurationItem;
}());
exports.default = FilterConfigurationItem;
